cumulus/polkadot-omni-node/lib/src/common/runtime.rs

```diff
@@ -62,7 +102,113 @@ pub trait RuntimeResolver {
+		}
+
+		Ok(Runtime::Omni(block_number, Consensus::Aura(AuraConsensusId::Sr25519)))
+	}
+}
+
+/// Inspect runtime metadata.
+pub struct MetadataInspector(subxt::Metadata);
+
+impl MetadataInspector {
+	/// Creates an instance of `MetadataInspector`.
+	pub fn new(chain_spec: &dyn ChainSpec) -> Result<MetadataInspector, sc_cli::Error> {
+		MetadataInspector::fetch_metadata(chain_spec).map(MetadataInspector)
+	}
+
+	/// Checks if pallet exists in runtime's metadata based on pallet name.
+	pub fn pallet_exists(&self, name: &str) -> bool {
+		self.0.pallet_by_name(name).is_some()
+	}
+
+	/// Get the configured runtime's block number type from `frame-system` pallet storage.
+	pub fn block_number(&self) -> Option<BlockNumber> {
+		let pallet_metadata = self.0.pallet_by_name(DEFAULT_FRAME_SYSTEM_PALLET_NAME);
+		pallet_metadata
+			.and_then(|inner| inner.storage())
+			.and_then(|inner| inner.entry_by_name("Number"))
+			.and_then(|number_ty| match number_ty.entry_type() {
+				StorageEntryType::Plain(ty_id) => Some(ty_id),
+				_ => None,
+			})
+			.and_then(|ty_id| self.0.types().resolve(*ty_id))
+			.and_then(|portable_type| BlockNumber::from_type_def(&portable_type.type_def))
+	}
+
+	/// Get the runtime metadata from a wasm blob existing in a chain spec.
+	pub fn fetch_metadata(chain_spec: &dyn ChainSpec) -> Result<subxt::Metadata, sc_cli::Error> {
+		let mut storage = chain_spec.build_storage()?;
+		let code_bytes = storage
+			.top
+			.remove(sp_storage::well_known_keys::CODE)
+			.ok_or("chain spec genesis does not contain code")?;
+		let opaque_metadata = fetch_latest_metadata_from_code_blob(
+			&WasmExecutor::<ParachainHostFunctions>::builder()
+				.with_allow_missing_host_functions(true)
+				.build(),
+			sp_runtime::Cow::Borrowed(code_bytes.as_slice()),
+		)
+		.map_err(|err| <String as Into<String>>::into(err.to_string()))?;
+
+		subxt::Metadata::decode(&mut (*opaque_metadata).as_slice()).map_err(Into::into)
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use crate::runtime::{
+		BlockNumber, MetadataInspector, DEFAULT_FRAME_SYSTEM_PALLET_NAME,
+		DEFAULT_PARACHAIN_SYSTEM_PALLET_NAME,
+	};
+	use codec::Decode;
+	use cumulus_client_service::ParachainHostFunctions;
+	use sc_executor::WasmExecutor;
+	use sc_runtime_utilities::fetch_latest_metadata_from_code_blob;
+
+	fn cumulus_test_runtime_metadata() -> subxt::Metadata {
+		let opaque_metadata = fetch_latest_metadata_from_code_blob(
+			&WasmExecutor::<ParachainHostFunctions>::builder()
+				.with_allow_missing_host_functions(true)
+				.build(),
+			sp_runtime::Cow::Borrowed(cumulus_test_runtime::WASM_BINARY.unwrap()),
+		)
+		.unwrap();
+
+		subxt::Metadata::decode(&mut (*opaque_metadata).as_slice()).unwrap()
+	}
+
+	#[test]
+	fn test_pallet_exists() {
+		let metadata_inspector = MetadataInspector(cumulus_test_runtime_metadata());
+		assert!(metadata_inspector.pallet_exists(DEFAULT_PARACHAIN_SYSTEM_PALLET_NAME));
+		assert!(metadata_inspector.pallet_exists(DEFAULT_FRAME_SYSTEM_PALLET_NAME));
+	}
```

Comment:
substrate_test_runtime does not have parachain system pallet, we could check if it work correctly with expected false result.

Reply:
We use `cumulus_test_runtime` actually, which defines the parachain system and frame system pallets: https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/test/runtime/src/lib.rs#L335. Asserting on a pallet name that doesn't exist fails as expected.